### ISSUE TITLE
First character of filename missing when file created locally

### ISSUE THREAD

**Issue by Lightslayer** [Azure/azure-storage-fuse#66]
Using blobfuse 0.2.2 package on Ubuntu 16.04 x86_64, blob files that are created using linux system command line tools such as `touch` are missing the first character in their filename when container storage refreshes on Azure portal

e.g. 

```
$ touch hello.world
$ ls -l
total 0
-rwxrwx--- 1 root        root         0 Jan  1  1970 ello.world
-rw-r--r-- 1 user        users        0 Jan 30 13:28 hello.world
```

From this it appears the file made by user root is the one that is uploaded to the blob container.

Is there any reason for this behaviour?

**Comment by seguler** [comment]
This is interesting. I don't see this behavior myself. What do you mean by when container storage refreshes on Azure portal ? Do you see ello.world file both in fuse and in the Portal ?

**Comment by Lightslayer** [comment]
Apologies I meant `ello.world` appears in both fuse and in the web portal, whereas `hello.world` seems to exist temporarily on fuse, but isn't propagated to the portal.

I mounted using the following command:
```
blobfuse /my_blob -o attr_timeout=240 -o entry_timeout=240 -o negative_timeout=120 --config-file=connection.cfg --tmp-path=/tmp/my-blob/
```

`/tmp/my-blob` was created using the same user performing the blobfuse mount.

I created `connection.cfg` following the wiki guide and substituting account, key and container as follows:
```
accountName <redacted>
accountKey <redacted>
containerName <redacted>
```

Note I also compiled the latest code from master and tried my locally built blobfuse, but with the same problem. As I seem to be the the only one experiencing this issue, maybe it is some incompatibility/instability with my system libs?

Not sure if it helps, but here is an excerpt from `apt list --installed`:

```
libfuse-dev/xenial-updates,now 2.9.4-1ubuntu3.1 amd64 [installed]
libfuse2/xenial-updates,now 2.9.4-1ubuntu3.1 amd64 [installed]
libgcrypt20/xenial-updates,xenial-security,now 1.6.5-2ubuntu0.3 amd64 [installed]
libgcrypt20-dev/xenial-updates,xenial-security,now 1.6.5-2ubuntu0.3 amd64 [installed]
libgnutls-dev/xenial-updates,now 3.4.10-4ubuntu1.4 amd64 [installed]
libgnutls-openssl27/xenial-updates,now 3.4.10-4ubuntu1.4 amd64 [installed]
libgnutls30/xenial-updates,now 3.4.10-4ubuntu1.4 amd64 [installed]
libgnutlsxx28/xenial-updates,now 3.4.10-4ubuntu1.4 amd64 [installed,automatic]
```

**Comment by seguler** [comment]
Thank you for the details. Do you mind enabling logs and sharing the logs here ? Edit blobfuse/blobfuse.h and set AZS_PRINT to 1 (it is 0) by default. And then rebuild running ./build.sh. This enables logs. You then eed to mount with -f option (blobfuse -f /path/to/mount ....). Reproduce the issue and let us know what you see on the console.

I suspect the issue is here: https://github.com/Azure/azure-storage-fuse/blob/master/blobfuse/fileapis.cpp#L490 
@asorrin-msft thoughts ? It seems we are blindly removing the first character here thinking it is a slash anyway ?

**Comment by asorrin-msft** [comment]
I think we do expect that the 'path' input into azs_truncate contains a leading slash.  I don't recall seeing any inconsistent behavior around this, but maybe there's something specific about your environment or your setup that doesn't include it?  I'm also not able to repro.

If we see that there's no leading slash passed in from FUSE in your case, we'll have to add some normalization to all the API's, they're all written assuming consistent behavior here.

Sorry about the terrible logging / output, I'll be working on that soon.

**Comment by seguler** [comment]
This is really weird. I have the exact same packages installed and used the exact same path you used to mount (/my_blob) and I can't get it to repro... A set of logs where you repro this issue would be greatly appreciated.

```
Ubuntu 16.04
libfuse-dev/xenial-updates,now 2.9.4-1ubuntu3.1 amd64 [installed]
libfuse2/xenial-updates,now 2.9.4-1ubuntu3.1 amd64 [installed]
libgcrypt20/xenial-updates,xenial-security,now 1.6.5-2ubuntu0.3 amd64 [installed]
libgcrypt20-dev/xenial-updates,xenial-security,now 1.6.5-2ubuntu0.3 amd64 [installed]
libgnutls-dev/xenial-updates,now 3.4.10-4ubuntu1.4 amd64 [installed,automatic]
libgnutls-openssl27/xenial-updates,now 3.4.10-4ubuntu1.4 amd64 [installed]
libgnutls28-dev/xenial-updates,now 3.4.10-4ubuntu1.4 amd64 [installed]
libgnutls30/xenial-updates,now 3.4.10-4ubuntu1.4 amd64 [installed]
libgnutlsxx28/xenial-updates,now 3.4.10-4ubuntu1.4 amd64 [installed,automatic]
```

**Comment by Lightslayer** [comment]
Apologies for the delay in getting back. Many thanks for the suggestions :-)

For testing I've setup a container with the following config:
```

accountName fusefiletruncationtest
accountKey Cb2r+MR7WkkHL+ntHnVBIoSqWGnVN0S3WQomXAG6KtlOylWFTnt47Er7H1AKeWc05F/5PrIZUjdW8N2MZxDNxw==
containerName public-test
```

Setting `AZS_PRINT` to 1 gave the following when mounting and creating a file:
```
~/azure-storage-fuse/build$ sudo mkdir -p /test_blob && sudo chown $USER:root /test_blob
~/azure-storage-fuse/build$ ~/azure-storage-fuse/build/blobfuse /test_blob -o attr_timeout=240 -o entry_timeout=240 -o negative_timeout=120 --config-file=blob_connection.cfg --tmp-path=/tmp/public-test-blob/
Making directory /tmp
Making directory /tmp/public-test-blob
Making directory /tmp/public-test-blob//root
~/azure-storage-fuse/build$ cd /test_blob/
/test_blob$ touch hello.world
/test_blob$ ls -l
total 0
-rwxrwx--- 1 root    root     0 Jan  1  1970 ello.world
-rw-r--r-- 1 user    users    0 Feb  1 10:38 hello.world
```

**Comment by seguler** [comment]
Could you rotate your account key ? You shared your root key here by mistake I suppose

**Comment by seguler** [comment]
Can you run blobfuse again with -f option ? The logs weren't showing up due to that.

You should mount it with:

blobfuse -f /test_blob -o attr_timeout=240 -o entry_timeout=240 -o negative_timeout=120 --config-file=blob_connection.cfg --tmp-path=/tmp/public-test-blob/

**Comment by seguler** [comment]
Ok - I reproduced the issue. It happens when you have a trailing slash in the path you provide in --tmp-path=/path/to/tmp/ 

As a workaround use --tmp-path=/tmp/public-test-blob

@asorrin-msft it seems we have a bug

**Comment by seguler** [comment]
This is now released in blobfuse 0.2.4
